gh-128481: Add missing documentation for `traceback.FrameSummary` (#128484)
python/cpython: Doc/library/traceback.rst

A :class:`!FrameSummary` object represents a single :ref:`frame <frame-objects>`
in a :ref:`traceback <traceback-objects>`.

> .. class:: FrameSummary(filename, lineno, end_lineno=None, name, lookup_line=True, locals=None, line=None, end_colno=None)

Reviewer: There's also an undocumented `colno` parameter that goes before `end_colno`. You can go ahead and document it too with this PR.

Author: Understood, I made the order of arguments of documentation consistent with the order of code, and it would like this
`   .. attribute:: FrameSummary.end_lineno

      The last line number of the source code for this frame.

   .. attribute:: FrameSummary.colno

      The column number of the source code for this frame.

   .. attribute:: FrameSummary.end_colno

      The last column number of the source code for this frame.`

Reviewer 2: > .. class:: FrameSummary(filename, lineno, end_lineno=None, name, lookup_line=True, locals=None, line=None, end_colno=None)

This is an illegal prototype. `end_lineno` cannot have a default None while `name` has no default. In addition, `lookup_line` are *keyword-only*:

```py
    def __init__(self, filename, lineno, name, *, lookup_line=True,
            locals=None, line=None,
            end_lineno=None, colno=None, end_colno=None):
```

Let's just copy the `__init__` signature for the docs.

> .. class:: FrameSummary(filename, lineno, name, lookup_line=True, locals=None, line=None, end_lineno=None, colno=None, end_colno=None)

Reviewer 2: ```suggestion
.. class:: FrameSummary(filename, lineno, name, *,\
                        lookup_line=True, locals=None,\
                        line=None, end_lineno=None, colno=None, end_colno=None)
```

Actually, the current (docs) signature as the code expect `lookup_line` and remaining parameters to be keyword only so let's also fix that here.

A :class:`!FrameSummary` object represents a single :ref:`frame <frame-objects>`
in a :ref:`traceback <traceback-objects>`.

.. class:: FrameSummary(filename, lineno, name, *,\

>                         lookup_line=True, locals=None, \

Reviewer 2: ```suggestion
                        lookup_line=True, locals=None,\
```

      trailing whitespace stripped.
      If the source is not available, it is ``None``.

>    .. attribute:: FrameSummary.end_lineno

Reviewer 2: Those new attributes can be ``None`` I think (just check the source code first). If so, this should be specified. Also, it'd be good to say whether it's a 0-indexed or 1-indexed base (I always forget)

Author: Based on the code description, those new attributes can be `None` that is right.
And those `number` attributes are 1-indexed base I think.

Reviewer 2: Actually, `end_lineno` is by default `lineno` if `end_lineno` is None. However, `colno` and `end_colno` can be stored as `None` (according to the source code). So by default it's `lineno` which itself can be None.

      trailing whitespace stripped.
      If the source is not available, it is ``None``.

   .. attribute:: FrameSummary.end_lineno

>       The last line number of the source code for this frame.
>       By default, it is ``None`` and is 1-index base.

Reviewer 2: ```suggestion
      The last line number of the source code for this frame.
      By default, it is ``None`` and indexation starts from 1.
```

   .. attribute:: FrameSummary.colno

      The column number of the source code for this frame.

>       By default, it is ``None`` and is 1-index base.

Reviewer 2: ```suggestion
      By default, it is ``None`` and indexation starts from 1.
```

   .. attribute:: FrameSummary.end_colno

      The last column number of the source code for this frame.

>       By default, it is ``None`` and is 1-index base.

Reviewer 2: ```suggestion
      By default, it is ``None`` and indexation starts from 1.
```

      trailing whitespace stripped.
      If the source is not available, it is ``None``.

>    .. attribute:: FrameSummary.line

Reviewer: Typo?
```suggestion
   .. attribute:: FrameSummary.end_colno
```

Author: Fixed it.

>    .. attribute:: FrameSummary.end_colno

Reviewer: There should be a newline here for consistency.

      trailing whitespace stripped.
      If the source is not available, it is ``None``.

   .. attribute:: FrameSummary.end_lineno

      The last line number of the source code for this frame.

>       By default, it is ``None`` and indexation starts from 1.

Reviewer 2: ```suggestion
      By default, it is set to ``lineno`` and indexation starts from 1.
```

   .. attribute:: FrameSummary.colno

      The column number of the source code for this frame.
      By default, it is ``None`` and indexation starts from 1.

   .. attribute:: FrameSummary.end_colno

      The last column number of the source code for this frame.
      By default, it is ``None`` and indexation starts from 1.

> 

Reviewer 2: ```suggestion

```

Author: What is the description of `end_colno` if modify as you suggested ? just leave it empty or delete entire documentation about it ?

Reviewer 2: Oups. I meant "add two blank lines" just before the examples section (we want two blank lines before ".. _traceback-example:")

      trailing whitespace stripped.
      If the source is not available, it is ``None``.

   .. attribute:: FrameSummary.end_lineno

      The last line number of the source code for this frame.
      By default, it is set to ``lineno`` and indexation starts from 1.

>    .. attribute:: FrameSummary.colno
> 
>       The column number of the source code for this frame.
>       By default, it is ``None`` and indexation starts from 1.
> 
>    .. attribute:: FrameSummary.end_colno
> 
>       The last column number of the source code for this frame.
>       By default, it is ``None`` and indexation starts from 1.

Reviewer 2: ```suggestion
   .. attribute:: FrameSummary.colno

      The column number of the source code for this frame.
      By default, it is ``None`` and indexation starts from 0.

   .. attribute:: FrameSummary.end_colno

      The last column number of the source code for this frame.
      By default, it is ``None`` and indexation starts from 0.
```

I'm sorry, for columns indexation starts from 0....
